courseware/api.py

```diff
@@ -1,48 +1,207 @@
 """Courseware API functions"""
-from datetime import datetime, timedelta
-from urllib.parse import urljoin
-import pytz
+from datetime import timedelta
+from urllib.parse import urljoin, urlparse, parse_qs
 import requests
 from rest_framework import status
 
 from django.conf import settings
+from django.db import transaction
+from django.shortcuts import reverse
 from oauth2_provider.models import Application, AccessToken
 from oauthlib.common import generate_token
 
-from courseware.models import CoursewareUser
+from authentication import api as auth_api
+from courseware.exceptions import OpenEdXOAuth2Error, CoursewareUserCreateError
+from courseware.models import CoursewareUser, OpenEdxApiAuth
 from courseware.constants import PLATFORM_EDX
+from mitxpro.utils import now_in_utc
 
 
 OPENEDX_REGISTER_USER_PATH = "/user_api/v1/account/registration/"
 OPENEDX_REQUEST_DEFAULTS = dict(country="US", honor_code=True)
 
+OPENEDX_SOCIAL_LOGIN_XPRO_PATH = "/auth/login/mitxpro-oauth2/?auth_entry=login"
+OPENEDX_OAUTH2_AUTHORIZE_PATH = "/oauth2/authorize"
+OPENEDX_OAUTH2_ACCESS_TOKEN_PATH = "/oauth2/access_token"
+OPENEDX_OAUTH2_SCOPES = ["read", "write"]
+OPENEDX_OAUTH2_ACCESS_TOKEN_PARAM = "code"
+OPENEDX_OAUTH2_ACCESS_TOKEN_EXPIRY_MARGIN_SECONDS = 10
+
+
+def edx_url(path):
+    """Returns the full url to the provided path"""
+    return urljoin(settings.OPENEDX_API_BASE_URL, path)
+
 
 def create_edx_user(user):
     """Makes a request to create an equivalent user in Open edX"""
     application = Application.objects.get(name=settings.OPENEDX_OAUTH_APP_NAME)
-    expiry_date = datetime.now(tz=pytz.utc) + timedelta(
-        hours=settings.OPENEDX_TOKEN_EXPIRES_HOURS
-    )
+    expiry_date = now_in_utc() + timedelta(hours=settings.OPENEDX_TOKEN_EXPIRES_HOURS)
     access_token, _ = AccessToken.objects.update_or_create(
         user=user,
         application=application,
         defaults=dict(token=generate_token(), expires=expiry_date),
     )
 
-    resp = requests.post(
-        urljoin(settings.OPENEDX_API_BASE_URL, OPENEDX_REGISTER_USER_PATH),
-        data=dict(
-            username=user.username,
-            email=user.email,
-            name=user.name,
-            provider=settings.MITXPRO_OAUTH_PROVIDER,
-            access_token=access_token.token,
-            **OPENEDX_REQUEST_DEFAULTS,
-        ),
-    )
-    # edX responds with 200 on success, not 201
-    if resp.status_code == status.HTTP_200_OK:
-        CoursewareUser.objects.create(
+    with transaction.atomic():
+        _, created = CoursewareUser.objects.select_for_update().get_or_create(
```

Comment:
I think this needs a few changes:
1. Change this query to `courseware_user, created = CoursewareUser.objects.select_for_update().get_or_create(user=user, platform=PLATFORM_EDX)`. If there are `CoursewareUser`s with `has_been_synced==False`, the query as it exists will try to create a new object, and that will fail due to the unique constraint
2. Change the return conditional from `if not created:` to `if not created or courseware_user.has_been_synced:`
3. Only set `courseware_user.has_been_synced` to `True` after the post request succeeds

Reply:
I agree on point 1, but the other changes don't align to how I thought this record was supposed to work, primarily: the existence of the record indicated whether it had been created and the sync field was meant to track updates (i.e. if we need to propagate profile updates).

[see discussion on PR that created this](https://github.com/mitodl/mitxpro/pull/81#discussion_r268221280)

Reply:
Ah, makes sense. In that case, we won't want to create the `CoursewareUser` at all until the post request succeeds, right?

Reply:
It's better to create the user and then roll it back if the request fails, that way if the db write fails after the request to the remote succeeds then we don't end up in a situation where we try to create the user again, which would fail because there's no way to rollback the edx write if the db write fails. Here's an example of the same strategy at work in discussions: https://github.com/mitodl/open-discussions/blob/master/channels/api.py#L674-L687

Reply:
Right, forgot about the transaction 👍